src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php

```diff
@@ -1447,26 +1448,33 @@ private function registerMessengerConfiguration(array $config, ContainerBuilder
 
         $loader->load('messenger.xml');
 
-        $container->setParameter('messenger.buses_names', array_keys($config['buses']));
+        $builtBuses = array();
         foreach ($config['buses'] as $name => $bus) {
-            $container->setDefinition($busId = 'messenger.bus.'.$name, (new Definition(MessageBus::class, array(array())))->addTag('messenger.bus')->setPublic(true));
+            $middlewares = array_map(function (array $middleware) use ($container) {
+                if ($container->has($messengerMiddlewareId = 'messenger.middleware.'.$middleware['service'])) {
+                    $middleware['service'] = $messengerMiddlewareId;
+                }
+
+                return new Reference($middleware['service']);
+            }, $bus['middlewares']);
 
-            if ($name == $config['default_bus']) {
+            $container->setDefinition($busId = 'messenger.bus.'.$name, (new Definition(MessageBus::class, array($middlewares)))->addTag('messenger.bus'));
+
+            if ($name === $config['default_bus']) {
                 $container->setAlias('message_bus', $busId);
                 $container->setAlias(MessageBusInterface::class, $busId);
             }
 
-            foreach ($bus['middlewares'] as $middleware) {
-                if (!$container->has($middleware['service'])) {
-                    $container->setDefinition($middleware['service'], new Definition($middleware['service']));
-                }
+            $container->setDefinition(
```

Comment:
so in user-land you need a compiler pass, find all buses by tag (`messenger.bus`), and decorate each one.

or is it better to decorate `call_message_handler` once?

Reply:
I'd say it depends on your use-case. Both approaches would work.